### PR TITLE
SONAR-6009 Each time the batch call the project referentials WS, a E/S query is executed for each rule

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-6009 Each time the batch call the project referentials WS, a E/S query is executed for each rule

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java [hunk 1]
```diff
@@ -88,8 +89,9 @@ public ProjectRepositories load(ProjectRepositoryQuery query) {
       if (module != null) {
         if (query.isPreview()) {
           // Scan permission is enough to analyze all projects but preview permission is limited to projects user can access
-          UserSession.get().checkComponentPermission(UserRole.USER, query.getModuleKey(),
-            "You're not authorized to access to project '" + module.name() + "', please contact your SonarQube administrator.");
+          if (!UserSession.get().hasProjectPermissionByUuid(UserRole.USER, module.projectUuid())) {
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java]: Just to understand here, what was the issue with ``checkComponentPermission`` vs ``hasProjectPermission``? Slower request?

File: server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java [hunk 2]
```diff
@@ -224,12 +226,13 @@ private QualityProfileDto getProfile(String languageKey, @Nullable String projec
 
   private void addActiveRules(ProjectRepositories ref) {
     for (org.sonar.batch.protocol.input.QProfile qProfile : ref.qProfiles()) {
-      for (ActiveRule activeRule : qProfileLoader.findActiveRulesByProfile(qProfile.key())) {
-        Rule rule = ruleService.getNonNullByKey(activeRule.key().ruleKey());
+      Map<RuleKey, ActiveRule> activeRules = activeRuleByRuleKey(qProfileLoader.findActiveRulesByProfile(qProfile.key()));
+      for (Rule rule : ruleService.search(new RuleQuery().setQProfileKey(qProfile.key()).setActivation(true), new QueryContext()).getHits()) {
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java]: Nice case of memory/resource tradeoff :+1:

**Comment by jblievremont** [comment]
LGTM!

File: server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java]: @jblievremont Indeed, checkComponentPermission will first search the project of the component, then check permission on the found project.
But direct check on hasProjectPermissionByUuid or checkProjectUuidPermission will only check permission on the project.

To sum up, there's one less SQL query when using XXXProjectUuisPermission !

File: server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/batch/ProjectRepositoryLoader.java]: :)

**Pull request merged**